### ISSUE TITLE
Link libcoreclr to libdl on macOS too

### ISSUE THREAD

**Issue by janvorli** [dotnet/runtime#32958]
We were never linking libcoreclr to libdl on macOS. It was fine due to the fact that the functions like dlopen are in fact defined in libSystem.dylib that we link to and the libdl.dylib is just a symbolic link to that library. However, since the libdl exists, we should link to that one too to be future proof in case the libdl starts being a real library and the dlopen and related methods get moved to that.
This was found based on issue #32852 where it seems that in several versions of macOS, trying to dlopen the libdl.dylib was failing unless the binary calling dlopen was explicitly linked to libdl (reported as a problem in a .NET core app, verified with basic C app).

**Comment by jkotas** [comment]
PInvoke of libdl is problematic on other Unix-like OSes too. IIRC, it has different name in some Linux distros. It is one of the reasons for why we have exposed the NativeLibrary API.

What would be the contract for PInvoke of libdl - are we going to make it always work, on all current and future Unix-like OSes? It basically means that we would be exposing PInvoke of libdl as a public surface. I think it would be better to point people to use the NativeLibrary API.

**Comment by janvorli** [comment]
It doesn't have a different name on Linux distros, the difference is on other Unix clones like FreeBSD where the dlopen is in the libc. 

I would not try to handle the libdl in a special way for pinvokes. As I've tried to say above, the only reason why I was thinking about adding the libdl to the list of libraries we link to was just to future proof our us against possible change in the future when macOS would make the libdl a real library instead of a symlink to libSystem and remove it from the libSystem.
However, when experimenting with it today, I've found that adding -ldl to the linked options (even with a simple C test app) doesn't have any effect and the binary still references only the libSystem. That would mean that such a change would only make the build process future proof, but not the binaries we generate. So it seems that there is not much need for this change now.

I am back talking to the people that created the issue #32852 trying to figure out how come adding -ldl had a positive effect on the problematic macOS 10.14.5.
